src/Sarif.Viewer.VisualStudio/Tags/SarifLocationTagger.cs

```diff
@@ -0,0 +1,553 @@
+            }
+        }
+
+        /// <summary>
+        /// Removes all tags.
+        /// </summary>
+        public static void RemoveAllTags()
+        {
+            IEnumerable<int> runsToRemove;
+            using (TagListLock.EnterReadLock())
+            {
+                runsToRemove = RunIdToSarifTags.Keys.ToList();
+            }
+
+            foreach (int runId in runsToRemove)
+            {
+                RemoveAllTagsForRun(runId);
+            }
+        }
+
+            /// <inheritdoc/>
+        public ISarifLocationTag AddTag(Region sourceRegion, TextSpan documentSpan, int runId, TextMarkerTag tag)
+        {
+            // Start an update so that even on a call (where the caller is not batching using Update themselves)
+            // that a "tags changed" event is fired to update VS's editor.
+            using (this.Update())
+            {
+                using (TagListLock.EnterWriteLock())
+                {
+                    SarifLocationTag existingSarifTag = this.sarifTags.FirstOrDefault(
+                        (sarifTag) =>
+                            sarifTag.SourceRegion.ValueEquals(sourceRegion));
+
+                    if (existingSarifTag != null)
+                    {
+                        return existingSarifTag;
+                    }
+
+                    IPersistentSpan persistentSpan = this.persistentSpanFactory.Create(
+                        this.textBuffer.CurrentSnapshot,
+                        startLine: documentSpan.iStartLine,
+                        startIndex: documentSpan.iStartIndex,
+                        endLine: documentSpan.iEndLine,
+                        endIndex: documentSpan.iEndIndex,
+                        SpanTrackingMode.EdgeInclusive);
+
+                    SarifLocationTag newSarifTag = new SarifLocationTag(
+                        persistentSpan,
+                        sourceRegion,
+                        runId,
+                        tag);
+
+                    this.sarifTags.Add(newSarifTag);
+                    newSarifTag.PropertyChanged += SarifTagPropertyChanged;
+
+                    if (RunIdToSarifTags.TryGetValue(runId, out List<SarifLocationTag> sarifTagsForRun))
+                    {
+                        sarifTagsForRun.Add(newSarifTag);
+                    }
+                    else
+                    {
+                        RunIdToSarifTags.Add(runId, new List<SarifLocationTag> { newSarifTag });
+                    }
+
+                    this.UpdateBatchSpan(newSarifTag.DocumentPersistentSpan.Span);
+
+                    return newSarifTag;
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public bool TryGetTag(Region sourceRegion, out ISarifLocationTag existingTag)
+        {
+            using (TagListLock.EnterReadLock())
+            {
+                existingTag = this.sarifTags.FirstOrDefault(sarifTag => sarifTag.SourceRegion.ValueEquals(sourceRegion));
```

Comment:
Why not Any?

Reply:
Any returns a bool and I need the reference to the tag.

---
In reply to: [481416588](https://github.com/microsoft/sarif-visualstudio-extension/pull/191#discussion_r481416588) [](ancestors = 481416588)